apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,441 @@
+	// +kubebuilder:validation:Optional
+	ResourceOverrideSnapshots []v1beta1.NamespacedName `json:"resourceOverrideSnapshots,omitempty"`
+
+	// ClusterResourceOverrides contains a list of applicable ClusterResourceOverride snapshot names associated with the
+	// selected resources.
+	// The list computed as the beginning of the update run and not updated during the update run.
+	// The list is empty if there are no cluster overrides associated with the selected resources.
+	// +kubebuilder:validation:Optional
+	ClusterResourceOverrideSnapshots []string `json:"clusterResourceOverrideSnapshots,omitempty"`
+
+	// CurrentStageIndex is the name of the stage that is in the middle of the update.
+	// If the update run is not started, the value is an empty string.
+	// If we are in the DeletionStage, the name is called "deletionStage".
+	// +kubebuilder:validation:Optional
+	CurrentUpdatingStage string `json:"currentUpdatingStage,omitempty"`
+
+	// Stages list the current status of each stage.
+	// The list is empty if the update run is not started.
+	// +kubebuilder:validation:Optional
+	StagesStatus []StageStatus `json:"stagesStatus,omitempty"`
+
+	// DeletionStageStatus list the current status of the deletion stage.
+	// +kubebuilder:validation:Optional
+	DeletionStageStatus StageStatus `json:"deletionStageStatus,omitempty"`
+
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	//
+	// Conditions is an array of current observed conditions for StagedUpdateRun.
+	// Known conditions are "RollingOut".
+	// +kubebuilder:validation:Optional
+	Conditions []metav1.Condition `json:"conditions"`
+}
+
+// StageStatus defines the status of the update run in a stage.
+type StageStatus struct {
+	// The name of the stage.
+	// +kubebuilder:validation:Required
+	StageName string `json:"stageName"`
+
+	// The status of the update run in the stage.
+	// +kubebuilder:validation:Required
+	Status UpdateStatus `json:"status"`
+
+	// The list of each cluster's update status in this stage.
+	// +kubebuilder:validation:Required
+	Clusters []ClusterUpdateStatus `json:"clusters"`
+
+	// The status of the post update tasks that are associated with current stage.
+	// +kubebuilder:validation:Optional
+	AfterStageTaskStatus []AfterStageTaskStatus `json:"currentPostStageTaskStatus,omitempty"`
+}
+
+// ClusterUpdateStatus defines the status of the update run on a cluster.
+type ClusterUpdateStatus struct {
+	// The name of the cluster.
+	// +kubebuilder:validation:Required
+	ClusterName string `json:"clusterName"`
+
+	// The status of the update run in the stage.
+	// +kubebuilder:validation:Required
+	Status UpdateStatus `json:"status"`
+}
+
+// UpdateStatus defines the status of a resource in an update run.
+type UpdateStatus struct {
+	// The status of the update on the cluster.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=NotStarted;Running;Completed;Failed;WaitingForApproval
```

Comment:
Waiting instead of WaitingForApproval? You could be waiting for timedwait or for approval